src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -576,10 +590,76 @@ public void onUpgradeComplete(final int newLevel)
           workOrder.isMirrored());
         this.setHeight(wrapper.getHeight());
         this.setCorners(corners.getFirst().getFirst(), corners.getFirst().getSecond(), corners.getSecond().getFirst(), corners.getSecond().getSecond());
+
+        // firework spawning possibly maybe
+
+        //ItemStack items = new ItemStack(f);
+
+
+
+        fireWorkCounter++;
+
+        if (fireWorkCounter % 2 == 1) {
+            final AxisAlignedBB realaabb = getTargetableArea(colony.getWorld());
+            final EntityFireworkRocket firework = new EntityFireworkRocket(colony.getWorld(), realaabb.maxX, realaabb.maxY, realaabb.maxZ, genFireworkItemStack());
+
+            colony.getWorld().spawnEntity(firework);
+            final EntityFireworkRocket fireworka = new EntityFireworkRocket(colony.getWorld(), realaabb.maxX, realaabb.maxY, realaabb.minZ, genFireworkItemStack());
+
+            colony.getWorld().spawnEntity(fireworka);
+            final EntityFireworkRocket fireworkb = new EntityFireworkRocket(colony.getWorld(), realaabb.minX, realaabb.maxY, realaabb.maxZ, genFireworkItemStack());
+
+            colony.getWorld().spawnEntity(fireworkb);
+            final EntityFireworkRocket fireworkc = new EntityFireworkRocket(colony.getWorld(), realaabb.minX, realaabb.maxY, realaabb.minZ, genFireworkItemStack());
+
+            colony.getWorld().spawnEntity(fireworkc);
+            System.out.println("POSITIONS: " + fireworkc.posX + ", " + fireworkc.posY + ", " + fireworkc.posZ);
+
+        }
+
+
+
+
+        //System.out.println("HEY THIS IS THE LOCATION:" + corners.getFirst().getFirst() + " " + corners.getFirst().getSecond() + " " + corners.getSecond().getFirst() + " " + corners.getSecond().getSecond());
+        //Minecraft.getMinecraft().player.playSound(SoundEvents.ENTITY_FIREWORK_LAUNCH, 1.0F, 1.0F);
+
         this.isBuilt = true;
     }
     //------------------------- Starting Required Tools/Item handling -------------------------//
+    private ItemStack genFireworkItemStack() {
```

Comment:
add javadoc
move method above the marker comment (as it doesn't seem to belong under `Starting Required Tools/Item handling`